From d2e19afa0f80a955b66ad133c4ebd5b1458bd8ce Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Rapha=C3=ABl=20Hertzog?= <hertzog@debian.org>
Date: Mon, 14 Nov 2016 11:46:09 +0100
Subject: [PATCH 1/2] Dpkg::Shlibs: add find_library_locations()

This new functions returns all paths/locations where the library has
been found. Where find_library() has some tricks to try to identify the
canonical version of the library, this one is much simpler: it scans all
directories from the library path and returns every place where it has
been found.
---
 scripts/Dpkg/Shlibs.pm | 27 +++++++++++++++++++++++++--
 1 file changed, 25 insertions(+), 2 deletions(-)

diff --git a/scripts/Dpkg/Shlibs.pm b/scripts/Dpkg/Shlibs.pm
index c6221c4..dc0276a 100644
--- a/scripts/Dpkg/Shlibs.pm
+++ b/scripts/Dpkg/Shlibs.pm
@@ -1,4 +1,4 @@
-# Copyright © 2007 Raphaël Hertzog <hertzog@debian.org>
+# Copyright © 2007,2016 Raphaël Hertzog <hertzog@debian.org>
 # Copyright © 2007-2008, 2012-2015 Guillem Jover <guillem@debian.org>
 #
 # This program is free software; you can redistribute it and/or modify
@@ -20,13 +20,14 @@ use strict;
 use warnings;
 use feature qw(state);
 
-our $VERSION = '0.02';
+our $VERSION = '0.03';
 our @EXPORT_OK = qw(
     blank_library_paths
     setup_library_paths
     get_library_paths
     add_library_dir
     find_library
+    find_library_locations
 );
 
 use Exporter qw(import);
@@ -174,4 +175,26 @@ sub find_library {
     return;
 }
 
+sub find_library_locations {
+    my ($lib, $rpath, $format, $root) = @_;
+
+    setup_library_paths() if not $librarypaths_init;
+
+    my @librarypaths = (@{$rpath}, @custom_librarypaths, @system_librarypaths);
+    my @locations;
+
+    $root //= '';
+    $root =~ s{/+$}{};
+    foreach my $dir (@librarypaths) {
+        my $checkdir = "$root$dir";
+        if (-e "$checkdir/$lib") {
+            my $libformat = Dpkg::Shlibs::Objdump::get_format("$checkdir/$lib");
+            if ($format eq $libformat) {
+                push @locations, canonpath("$checkdir/$lib");
+            }
+        }
+    }
+    return @locations;
+}
+
 1;
-- 
2.10.2

